### PR TITLE
[jit] Switch import/export to python printing

### PR DESCRIPTION
**Pull request by zdevito**
Stacked on https://github.com/pytorch/pytorch/pull/14378, only look at the last commit.

This changes the way methods are defined in TorchScript archives to use
PythonPrint rather than ONNX protobufs.

It also updates torch.proto to directly document the tensor data
structure actually being serialized.

Notes:
* because PythonPrint prints all the methods at once per module, this
  removes MethodDef in favor of a single torchscript_area and a separate
  caffe2_graphs entry. Note that NetDef's already have method names,
  so there is no need or a separate method name entry.
* This switches cpp/pickle area to RecordRef (references to a file in
  the container format) since it is possible the data in these arenas
  may be large and not suited to json ouput.
* Removes 'annotations' -- annotations should be re-added on the first
  commit that actually has a practical use for them. In the current state
  it is unlikely they are representing the right information.
* Some expect files have changed because PythonPrint is preserving more
  debug name information for parameter names.
* MethodEncoder (the ONNX output format) has been deleted. There is still
  some cleanup possible combining EncoderBase and GraphEncode now that there
  is only a single pathway using EncoderBase.
* This incorporates the changes from #14397
  to define TensorDef

### CODE REVIEW

**Review by dzhulgakov — COMMENTED** [review]
cc @wanchaol who's been working on undefined tensor serialization

File: torch/csrc/jit/export.cpp [hunk 1]
```diff
@@ -752,157 +516,79 @@ void ScriptModuleSerializer::serialize(const script::Module& module) {
...
+  return tensor_table_.size() - 1;
 }
 
 void ScriptModuleSerializer::convertAndWriteTensor(
     const at::Tensor& tensor,
-    caffe2::TensorProto* tensor_proto) {
-  auto tensor_it = tensorTable_.find(&tensor);
-  AT_ASSERT(tensor_it != tensorTable_.end());
-  tensor_proto->set_name(c10::to_string(tensor_it->second));
+    torch::TensorDef* tensor_proto,
```
Comment (houseroad) [inline torch/csrc/jit/export.cpp]: Nit: rename tensor_proto to tensor_def

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -752,157 +516,79 @@ void ScriptModuleSerializer::serialize(const script::Module& module) {
...
     }
-    external_data->set_record_id(c10::to_string(record_id));
-    storageMap_[key] = record_id;
-  } else {
-    external_data->set_record_id(c10::to_string(storage_it->second));
+    uint64_t record_id = writer_.writeRecord(storage_tensor.storage().data(), record_size);
+    storage_it = storageMap.insert({key, record_id}).first;
   }
+
+  auto data = tensor_proto->mutable_data();
```
Comment (houseroad) [inline torch/csrc/jit/export.cpp]: nit: auto* to make it clear it's a pointer

File: torch/csrc/jit/import.cpp [hunk 3]
```diff
@@ -400,111 +97,68 @@ void ScriptModuleDeserializer::deserialize(ModuleLookup module_lookup) {
...
-  for (int i = 0; i < external_data.strides_size(); ++i) {
-    strides.push_back(external_data.strides(i));
+void ScriptModuleDeserializer::loadTensorTable(torch::ModelDef* model_def) {
+  std::unordered_map<uint64_t, at::Storage> storageMap;
+  for(const torch::TensorDef& tensor : model_def->tensors()) {
+    tensor_table_.emplace_back(loadTensor(tensor, storageMap));
   }
+}
+
+at::Tensor ScriptModuleDeserializer::loadTensor(const torch::TensorDef& tensor_proto,
```
Comment (houseroad) [inline torch/csrc/jit/import.cpp]: Nit: tensor_proto => tensor_def

File: torch/csrc/jit/passes/python_print.cpp [hunk 4]
```diff
@@ -193,7 +193,7 @@ struct PythonPrintPass {
   // constants are written to this table, and given then named CONSTANTS.cN
   // where N is the index into this table.
 
-  std::vector<at::Tensor> tensor_constants;
+  std::vector<at::Tensor>& tensor_table_;
```
Comment (houseroad) [inline torch/csrc/jit/passes/python_print.cpp]: Nit: shall we turn tensor_table_ to a pointer type (i.e., std::vector<at::Tensor>*)?
Having a reference type as a member looks not very appropriate to me.

**Review by houseroad — COMMENTED** [review]
For the fields in the proto, we probably can reuse the field numbers, since 1) there are no serialized models, 2) even there are serialized models in old format, we don't keep the ability to load them, doesn't make sense to reserve the fields before the stable release.

Some inline nits

**Comment by houseroad** [comment]
@wanchaol @dzhulgakov is data_type=UNDEFINED enough for serializing undefined tensor?

**Comment by wanchaol** [comment]
> @wanchaol @dzhulgakov is data_type=UNDEFINED enough for serializing undefined tensor?
> 
> Update: using special value (or empty) for tensor_id should be enough.

Yes I choose to use empty `tensor_id` field to indicate an undefined tensor in #14379

File: caffe2/proto/torch.proto [hunk 5]
```diff
@@ -4,52 +4,57 @@ import "caffe2/proto/caffe2.proto";
...
+  optional int64 size = 2;
+}
+
+message TensorDef {
+  repeated int64 dims = 1;
+  optional int64 offset = 2;
+  repeated int64 strides = 3;
   // whether we compute the gradient for the parameter
-  optional bool require_gradient = 1;
+  optional bool requires_grad = 4;
```
[the remark this answers is not in the document]
Comment (zdevito) [reply caffe2/proto/torch.proto]: I think so -- when torch::Variable no longer exists, we will need to construct a Tensor that has requires_grad set correctly, so it makes more sense to do it directly on the tensor. It is not an issue for constants, they will always get serialized with requires_grad=False.

**Comment by zdevito** [comment]
@wanchaol  -- If we get rid of prim::Undefined and universally use prim::None, then we will never need to serialize an undefined Tensor, right?

**Comment by wanchaol** [comment]
> @wanchaol -- If we get rid of prim::Undefined and universally use prim::None, then we will never need to serialize an undefined Tensor, right?

@zdevito Partially, if we get rid of prim::Undefined, we don't need to serialize undefined tensors in the nodes of graph body, but there will still be cases that we need to serialize undefined tensors. The case we need it is when ScriptModule/nn.Module parameter can accept a NoneType, an example is in [_BatchNorm](https://github.com/pytorch/pytorch/blob/3ff70712c29f7bc61f3861cae1ce6ce3122a30b2/torch/nn/modules/batchnorm.py#L37), `weight` and `bias` can be initialized as tensor or None depending on `self.affine`. The None here need to be interpreted as undefined tensor (otherwise we cannot handle it since it also could not appear in ScriptModule's `__constants__`),  `ScriptModule` parameters will be turned into graph input nodes, and those nodes (including the undefined tensor node) need to be serialized/deserialized as well.

File: torch/csrc/jit/passes/python_print.cpp (hunk 4 above)
Comment (zdevito) [reply torch/csrc/jit/passes/python_print.cpp]: I don't see a big reason to change it, we already use an ostream& in this class as well. The class is really just being used a a closure that holds common fields across the recursive set of export methods.

**Review by suo — APPROVED** [review]
